app/services/search/postgres/reading_list.rb

```diff
@@ -16,36 +16,93 @@ def self.search_documents(user, term: nil, status: nil, tags: nil, tags_mode: :a
+        # NOTE: [@rhymes] I'm not sure how to combine `pg_search`'s multisearch
+        # with the `.tagged_with()` scope. I could hack it by manually filtering articles based on their
+        # `cached_tag_list` but then we'd break pagination, as the filtering needs to happen before
+        # results are split in pages
+        relation = Search::Multisearch.call(term)
+          .with_pg_search_highlight
+          .with_pg_search_rank
+          .includes(searchable: %i[user tags])
+          .where(searchable_type: "Article", searchable_id: article_ids)
+
+        # NOTE: acts-as-taggable-on generates SQL at runtime to join `taggings`, `tags` and `articles`
+        # we can't directly attach it to `PgSearch::Document` as we're indeed trying to filter articles whose
+        # tags appear in the list. For this implementation I decided to do a 2-step algorithm: search and filter.
+        # Rather than trying to build a monster SQL query that joins `pg_search_documents`, `taggings`, `tags` and
+        # `articles` (it can be done manually but we kinda lose acts-as-taggable-on dynamic features) I decided to
+        # re-use `ActsAsTaggableOn::Taggable::TaggedWithQuery` to filter on articles that have previously matched
+        # by the FTS search.
+        # Keeping in mind that we're searching inside a reading list (a subset of saved articles), this should be fast
+        # enough.
+        if tags.present?
+          # `tagged_with` returns article objects, we tell it to find tagged articles within those returned
+          # by the search
+          relation = tagged_with(relation.map(&:searchable_id), tags, tagged_with_any)
+          relation.page(page).per(per_page)
+
+          # NOTE: commented this out for now as it's only needed if we actually want to match
+          # `pg_search_highlight`/`pg_search_highlight` with the results, but they aren't currently used
+
+          # filtered_articles = tagged_articles.map do |article|
+          #   doc = search_documents[article.id]
+          #   next unless doc
 
+          #   article.search_highlight = doc.pg_search_highlight
+          #   article.search_rank = doc.pg_search_rank
+          #   article
+          # end
+
+          # Kaminari.paginate_array(filtered_articles).page(page).per(per_page)
+        else
+          relation = relation.page(page).per(per_page)
+
+          # extract Article objects from `PgSearch::Document` objects
+          relation.map do |doc|
+            article = doc.searchable
+            article.search_highlight = doc.pg_search_highlight
+            article.search_rank = doc.pg_search_rank
+            article
+          end
+        end
+      end
+      private_class_method :full_text_search
+
+      # borrowed from https://github.com/mbleigh/acts-as-taggable-on/blob/47da5036dea61cb971bfaf72de5fa93c85255307/lib/acts_as_taggable_on/taggable/core.rb#L110
+      def self.tagged_with(article_ids, tags, tagged_with_any)
```

Comment:
Could we potentially do this by changing the first part of this method to use `super`? I think

```ruby
def self.tagged_with(article_ids, tags, tagged_with_any)
  relation = super(tags, { any: tagged_with_any})
  relation.where(id: article_ids)
end
```

Reply:
Can't use `super` as the class doesn't include the `acts-as-taggable-on` module BUT we can just call it directly :D 

```ruby
      def self.tagged_with(article_ids, tags, tagged_with_any)
        relation = Article.tagged_with(tags, any: tagged_with_any)
        relation.where(id: article_ids)
      end
      private_class_method :tagged_with
```

Reply:
Ah, yes, that works, thanks!